### PR TITLE
Support local actions that output more than 65535 chars

### PR DESCRIPTION
**Pull request by manasdk**
- Using Popen.wait() with STDOUT=subprocess.PIPE causes a deadlocal and it is
  described as a warning in the docs as well. Basically, the child process blocks
  on the buffer from being empty and the spawning process is wait for the child to
  quite prior to reading - so deadlock.
- The solution is to use Popen.communicate(). Using this works great and large
  data upto 10^6 chars can also be read. Rest of the system starts struggling but
  that is not in scope of this commit.
- Using communicate leads to challenges with kill on a timeout. The timeout is placed
  to avoid bad action from taking down the runners. The only way to cleanly handle the
  kill of all the process is to use pkill. The specific way in which the shell command
  is constructed implies that there is no child-parent relation. Therefore all procs
  are put in a single process group and the process group is killed. This requires
  use of pkill. We clearly don't plan to run this code on windows.

### CODE REVIEW

**Comment by dzimine** [comment]
Beyond this specific PR: 

Do you consider switching to result tracker for these verbouse-outputting actions? This brings back 'async actions' :)

My quick thoughts: 1)  it's difficult to optimize for conflicting goals; fast action / small output and long action / large output (I assume fast task / large output is a corner case but if we need to separate, we separate by output size) 2) system can not predict size of output, but the user can in most practical cases. 3) each type of actions (each runner) can have both short and long running actions 4) generic actions -  run-local / run-remote, or HTTP etc - may be both fast/small and long/large, so 'sync/async' behavior may be a runner parameter.

I don't see this as immediate priority unless there's a use case from @jfryman or @DoriftoShoes, just want to plant the through for you guys to elaborate when we have more time.

**Comment by manasdk** [comment]
> Do you consider switching to result tracker for these verbouse-outputting actions? This brings back 'async actions' :)

Yes. Async actions is a real usecase which I believe we should target. Changes will be needed in the Querier mechanism and also a notion of per-action querier. So will take sometime to get there but that is the correct direction.

**Comment by jfryman** [comment]
I certainly do not need this right now, but it will come up. A few use cases I can imagine:

* Hostname queries
* responses from ChatOps commands (that are Shell in nature)

**Comment by manasdk** [comment]
> I don't see this as immediate priority

By default we should assume that all action generate an overwhelming amount of data and take infinite time. If we satisfy this requirement then we can satisfy pretty much all requirements like fast-low, fast-high etc. The only reason not to have a generic solution is we can establish that the generic solution is too slow for the 80% use case. On the priority I agree that we don't have to worry about it just yet aka we other more pressing concerns :)

This PR is a step in the direction of handling large-ish data without really truly solving the problem entirely. However, fixing the deadlock is required so we have to push through.
At some point we will have to come back to solving this but perhaps in a different context and most likely with greater clarity.

**Comment by manasdk** [comment]
> responses from ChatOps commands (that are Shell in nature)
What is special about ChatOps? Does the source of the command have any special implications?

**Comment by jfryman** [comment]
> What is special about ChatOps? Does the source of the command have any special implications?

I cannot say for sure that the source matters... but there were a ton of shell commands that we executed that outputted many, many lines of text and would be truncated. (an `rsync` diff action is one that immediately comes to mind)

**Comment by manasdk** [comment]
>  but there were a ton of shell commands that we executed that outputted many, many lines of text and would be truncated.

In context of this PR the output will not be truncated when we read it from the actual action process. However, I can imagine that the ChatOps, UI or CLI and a client could choose to render only a part of the output as per their requirements. Therefore, 
1. The backend needs to deal with large outputs and capture them fully without dying.
2. Perhaps the API needs to be provide control of how much of the result to return so that an unnecessary amount of data is not transferred to clients. This should help render large data - this is an example of features we would need to build to deal with large outputs.

File: st2actions/tests/integration/test_localrunner.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+from unittest2 import TestCase
+from st2actions.container.service import RunnerContainerService
+from st2actions.runners import localrunner
+from st2common.constants import action as action_constants
+from st2tests.fixturesloader import FixturesLoader
+
+
+class TestLocalShellRunner(TestCase):
+
+    fixtures_loader = FixturesLoader()
```
Comment (lakshmi-kannan) [inline st2actions/tests/integration/test_localrunner.py]: I kinda hate doing this. I think DbTestCase should come with a fixtures loader. Whaddya think? We can make this change in another PR.

Comment (manasdk) [inline st2actions/tests/integration/test_localrunner.py]: Yeah. That would work great. In this case I was not really loading into DB so would not have use DbTestCase anyway.

**Comment by lakshmi-kannan** [comment]
LGTM. Good test cases. CC: @Kami

**Comment by Kami** [comment]
Good catch and nice tests.

IIRC, I originally didn't use `.communicate()` because it was blocking. Looking at the code again though (https://github.com/eventlet/eventlet/blob/master/eventlet/green/subprocess.py#L79), it seems like this should only be a problem on Windows so we should be fine.

File: st2actions/st2actions/runners/localrunner.py [hunk 2]
```diff
@@ -106,21 +109,40 @@ def run(self, action_parameters):
...
+                error_holder['error'] = 'Action failed to complete in %s seconds' % (self._timeout)
+                # Action has timed out, kill the process and propagate the error. The process
+                # is started as sudo -u {{system_user}} -- bash -c {{command}}. Introduction of the
+                # bash means that multiple independent processes are spawned without them being
+                # children of the process we have access to and this requires use of pkill.
+                # Ideally os.killpg should have done the trick but for some reason that failed.
+                # Note: pkill will set the returncode to 143 so we don't need to explicitly set
+                # it to some non-zero value.
+                try:
+                    killcommand = shlex.split('sudo pkill -TERM -s %s' % process.pid)
```
Comment (Kami) [inline st2actions/st2actions/runners/localrunner.py]: Since we control the input here it might be a bit better and safer to just use a list here instead of `shlex.split` (`args = ['sudo', 'pkill', '-TERM", '-s', process.pid]`).

Using `shlex.split` means traditional shell injection should not be possible, but if a value of `process.pid` can be controlled, it can be made to pass multiple arguments to `pkill` which is not possible when using a list.

Comment (manasdk) [inline st2actions/st2actions/runners/localrunner.py]: That makes sense. Will open a separate PR to bring in this change.
